### PR TITLE
Added deep merging support to rjsone

### PR DESCRIPTION
**Pull request by ychen-atlassian**
Added an option `-d` which uses mergo to run a deep merge of passed in contexts.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -287,8 +312,8 @@ func buildFunction(commandString string, rawOutput, rawInput bool) interface{} {
 	if rawInput && rawOutput {
 		f = func(args []interface{}, stdin string) (string, error) {
 			stringArgs, err := castToStrings(args)
-			commandArray = append(commandArray, stringArgs...)
-			command := exec.Command(commandArray[0], commandArray[1:]...)
+			extendedCommandArray := append(commandArray, stringArgs...)
```
Comment (ychen-atlassian) [inline main.go:315]: I found a bug here. The returned function `f` is a closure over `commandArray`. We evaluate `buildFunction` once for the arguments passed in, but we invoke `f` for every invocation.

This means every time you invoke `f` you end up appending to commandArray. So your command ends up being this gigantic aggregate of all your arguments.

Obviously I fixed this by declaring the command array as a new array inside `f`.

File: testdata/complex/run.sh [hunk 2]
```diff
@@ -1,3 +1,3 @@
 #!/bin/sh
 
-rjsone -y -t template.yaml b64decode::--'base64 -d' b64encode::--base64 rawparam::+hello context1.yaml context2.yaml foobar:named.yaml text::input.txt list:.. context1.yaml context2.yaml list::.. input.txt withfilename::... input.txt
```
Comment (ychen-atlassian) [inline testdata/complex/run.sh:3]: I replaced this with something more friendly to OSX and every other environment that doesn't have base64. Usually `tr` is more common than `base64`.

File: test.sh [hunk 3]
```diff
@@ -2,16 +2,35 @@
 
 set -eu
 
+xreadlink() {
```
Comment (ychen-atlassian) [inline test.sh:5]: Again `readlink -f` doesn't work on OSX so I wrote my own.

File: test.sh [hunk 4]
```diff
@@ -22,7 +41,7 @@ run() {
 PASS=0
 FAIL=0
 
-if [ "$1" = "-g" ]; then
+if [ "${1-}" = "-g" ]; then
```
Comment (ychen-atlassian) [inline test.sh:44]: Another bugfix - if you gave the script no arguments it vomits. I mean not really a big deal but meh,

File: main.go (hunk 1 above)
Comment (wryun) [reply main.go:315]: Err, whoops. Thanks!

File: test.sh [hunk 5]
```diff
@@ -2,16 +2,35 @@
...
+
+   while [ -L "$node" ]
+   do
+       node=`readlink $node`
+
+       cd `dirname $node`
+       node=`basename $node`
+   done
+
+   echo `pwd`/$node
```
Comment (wryun) [inline test.sh:21]: Heh, this won't work if there are IFS characters other than space in the path... not sure if any filesystems support this...

File: testdata/merge/template.yaml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+should_be_overridden: {$eval: root.subkey1}
```
Comment (wryun) [inline testdata/merge/template.yaml:1]: no real need to use eval here, just FYI (can just use string interp).

**Comment by wryun** [comment]
FYI, when you update the comment in main.go it's good to copy/paste into README.md (for github + same file is used for the website).

**Pull request merged**